app/services/immediate_email_generator_service.rb

```diff
@@ -1,27 +1,20 @@
-class ImmediateEmailGenerationWorker
-  include Sidekiq::Worker
-
-  sidekiq_options queue: :email_generation_immediate
-
-  LOCK_NAME = "immediate_email_generation_worker".freeze
+module ImmediateEmailGeneratorService
```

Comment:
This mixin used for inheritance doesn't seem an ideal approach to me here, at first I wondered should you just use standard inheritance as that seems less surprising than the module choice for it, though after having a think I don't think you need it at all.

I don't think you gain much by having the lock in a separate method and it might be better to ditch `ensure_only_running_once` to just use the actual lock in both things. Like Tom I do wonder if we can advisory the need for an advisory lock. I'd agree with your comment that we need some form of protection but I'd hope we could do it with db transactions and constraints rather than explicit locks.

With the generate_emails method I think this should have distinct variations for content change and messages. At the moment it does 5 tasks where only 3 are relevant for ContentChanges and Messages respectively so there's a bunch of shared methods that are only relevant on the subclasses.

There's also likely quite a lot of redundant stuff going on. As we only have one Content Change or one Message we shouldn't need a cache of them and it may be that we want to combine looking up by subscribers and their subscriptions given they'd be for the same item.